### PR TITLE
core: split external script unit test and main code in diff file

### LINKED ISSUE #9588: split unit test and original code of python script in different file
It's better to split the unit test and original code of the [python](https://github.com/rook/rook/blob/master/deploy/examples/create-external-cluster-resources.py) script into a different file.

### PR DESCRIPTION
**Pull request by subhamkrai**
split the external script main code and unit test in a different files.

Signed-off-by: subhamkrai <srai@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #9588 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Review by parth-gr — CHANGES_REQUESTED** [review]
https://github.com/rook/rook/blob/7dc89a49fa9b86f253153350e841916bf4a7f4f1/deploy/examples/create-external-cluster-resources.py#L84
All these definitions/ declarations belong to the unit test too.

**Comment by subhamkrai** [comment]
> https://github.com/rook/rook/blob/7dc89a49fa9b86f253153350e841916bf4a7f4f1/deploy/examples/create-external-cluster-resources.py#L84
> 
> All these definitions/ declarations belong to the unit test too.

actually, not only these pieces but the whole class https://github.com/rook/rook/blob/7dc89a49fa9b86f253153350e841916bf4a7f4f1/deploy/examples/create-external-cluster-resources.py#L68 but the way we have script today is main code uses that class `DummyRados` and we discussed(I and @aruniiird ) the same and decided to leave these parts in the main file only.

**Review by leseb — CHANGES_REQUESTED** [review]
LGTM but let's wait for https://github.com/rook/rook/pull/9609 first

**Review by leseb — APPROVED** [review]

**Comment by parth-gr** [comment]
> but the way we have script today is main code uses that class `DummyRados` and we discussed(I and @aruniiird ) the same and decided to leave these parts in the main file only.

I see only we use `DummyRados` at 2 places, line 442 323, and it's used only for checking if the cluster type is from unit test or not,
First of all, we should not do that, changing the function just for the sake of a unit test.

In a follow-up PR or in this, can you try removing those instances and checking what impact it gives to unit test,  I think these will run successfully,
I am too much emphasising into it because I feel like the class `DummyRados`  is part of the unit test and should be abstract to the actual code.
Thankx

**Comment by subhamkrai** [comment]
> > but the way we have script today is main code uses that class `DummyRados` and we discussed(I and @aruniiird ) the same and decided to leave these parts in the main file only.
> 
> I see only we use `DummyRados` at 2 places, line 442 323, and it's used only for checking if the cluster type is from unit test or not, First of all, we should not do that, changing the function just for the sake of a unit test.
> 
This is something that is not introduced in my PR, it is there already
> In a follow-up PR or in this, can you try removing those instances and checking what impact it gives to unit test, I think these will run successfully, I am too much emphasising into it because I feel like the class `DummyRados` is part of the unit test and should be abstract to the actual code. Thankx

I agree,  `DummyRados` should be part of the unit test only and I already mentioned that earlier and discussed it with @aruniiird 
We can discuss more in this.

File: deploy/examples/create-external-cluster-resources-tests.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+'''
+
+import unittest
+import importlib
+import sys
+import os
+import sys
```
Comment (parth-gr) [inline deploy/examples/create-external-cluster-resources-tests.py:18]: sys imported 2 times line 16 and 18

File: deploy/examples/create-external-cluster-resources-tests.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+import json
+
+file_path = os.path.dirname(__file__)
+sys.path.append(file_path)
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply deploy/examples/create-external-cluster-resources-tests.py:22]: okay, will ask you offline.

**Review by parth-gr — APPROVED** [review]

**Review by parth-gr — APPROVED** [review]

**Comment by aruniiird** [comment]
> > > but the way we have script today is main code uses that class `DummyRados` and we discussed(I and @aruniiird ) the same and decided to leave these parts in the main file only.
> > 
> > 
> > I see only we use `DummyRados` at 2 places, line 442 323, and it's used only for checking if the cluster type is from unit test or not, First of all, we should not do that, changing the function just for the sake of a unit test.
> 
> This is something that is not introduced in my PR, it is there already
> 
> > In a follow-up PR or in this, can you try removing those instances and checking what impact it gives to unit test, I think these will run successfully, I am too much emphasising into it because I feel like the class `DummyRados` is part of the unit test and should be abstract to the actual code. Thankx
> 
> I agree, `DummyRados` should be part of the unit test only and I already mentioned that earlier and discussed it with @aruniiird We can discuss more in this.

About the DummyRados, there are some difficulties in removing the (DummyRados) class from the `create-external-cluster-resources.py` file.
a. if we move the DummyRados definition to the test file, without removing the usage in the `create-external-cluster-resources.py` file we will create a cyclic dependency.
b. if we remove all the references of DummyRados from the `create-external-cluster-resources.py` file and then move it to the test file, then it should be fine, but we will face difficulties in running the test.

The clean way out will be to take DummyRados class out to it's own `py` file and import it accordingly.

I agree to @parth-gr in the sentiments that we should not include/introduce DummyRados into the actual code. But we need to find a workaround for it...

Thanks

**Comment by subhamkrai** [comment]
> > > > but the way we have script today is main code uses that class `DummyRados` and we discussed(I and @aruniiird ) the same and decided to leave these parts in the main file only.
> > > 
> > > 
> > > I see only we use `DummyRados` at 2 places, line 442 323, and it's used only for checking if the cluster type is from unit test or not, First of all, we should not do that, changing the function just for the sake of a unit test.
> > 
> > 
> > This is something that is not introduced in my PR, it is there already
> > > In a follow-up PR or in this, can you try removing those instances and checking what impact it gives to unit test, I think these will run successfully, I am too much emphasising into it because I feel like the class `DummyRados` is part of the unit test and should be abstract to the actual code. Thankx
> > 
> > 
> > I agree, `DummyRados` should be part of the unit test only and I already mentioned that earlier and discussed it with @aruniiird We can discuss more in this.
> 
> About the DummyRados, there are some difficulties in removing the (DummyRados) class from the `create-external-cluster-resources.py` file. a. if we move the DummyRados definition to the test file, without removing the usage in the `create-external-cluster-resources.py` file we will create a cyclic dependency. b. if we remove all the references of DummyRados from the `create-external-cluster-resources.py` file and then move it to the test file, then it should be fine, but we will face difficulties in running the test.
> 
> The clean way out will be to take DummyRados class out to it's own `py` file and import it accordingly.
> 
> I agree to @parth-gr in the sentiments that we should not include/introduce DummyRados into the actual code. But we need to find a workaround for it...
>

are we good for this pr or do we need to think about some workaround and include it in this PR? 

> Thanks

**Comment by aruniiird** [comment]
> > > > > but the way we have script today is main code uses that class `DummyRados` and we discussed(I and @aruniiird ) the same and decided to leave these parts in the main file only.
> > > > 
> > > > 
> > > > I see only we use `DummyRados` at 2 places, line 442 323, and it's used only for checking if the cluster type is from unit test or not, First of all, we should not do that, changing the function just for the sake of a unit test.
> > > 
> > > 
> > > This is something that is not introduced in my PR, it is there already
> > > > In a follow-up PR or in this, can you try removing those instances and checking what impact it gives to unit test, I think these will run successfully, I am too much emphasising into it because I feel like the class `DummyRados` is part of the unit test and should be abstract to the actual code. Thankx
> > > 
> > > 
> > > I agree, `DummyRados` should be part of the unit test only and I already mentioned that earlier and discussed it with @aruniiird We can discuss more in this.
> > 
> > 
> > About the DummyRados, there are some difficulties in removing the (DummyRados) class from the `create-external-cluster-resources.py` file. a. if we move the DummyRados definition to the test file, without removing the usage in the `create-external-cluster-resources.py` file we will create a cyclic dependency. b. if we remove all the references of DummyRados from the `create-external-cluster-resources.py` file and then move it to the test file, then it should be fine, but we will face difficulties in running the test.
> > The clean way out will be to take DummyRados class out to it's own `py` file and import it accordingly.
> > I agree to @parth-gr in the sentiments that we should not include/introduce DummyRados into the actual code. But we need to find a workaround for it...
> 
> are we good for this pr or do we need to think about some workaround and include it in this PR?
> 
> > Thanks

Changing the `DummyRados` (IMHO) to a different file and changing the references can be taken up as a different refactoring PR. But I've a small concern regarding `importlib` package, which is only there in 3.x python version. Is there any way we can make this backward compatible?

If are not concerned about the 2.7.x python versions, then LGTM from my end...

**Comment by subhamkrai** [comment]
> > > > > > but the way we have script today is main code uses that class `DummyRados` and we discussed(I and @aruniiird ) the same and decided to leave these parts in the main file only.
> > > > > 
> > > > > 
> > > > > I see only we use `DummyRados` at 2 places, line 442 323, and it's used only for checking if the cluster type is from unit test or not, First of all, we should not do that, changing the function just for the sake of a unit test.
> > > > 
> > > > 
> > > > This is something that is not introduced in my PR, it is there already
> > > > > In a follow-up PR or in this, can you try removing those instances and checking what impact it gives to unit test, I think these will run successfully, I am too much emphasising into it because I feel like the class `DummyRados` is part of the unit test and should be abstract to the actual code. Thankx
> > > > 
> > > > 
> > > > I agree, `DummyRados` should be part of the unit test only and I already mentioned that earlier and discussed it with @aruniiird We can discuss more in this.
> > > 
> > > 
> > > About the DummyRados, there are some difficulties in removing the (DummyRados) class from the `create-external-cluster-resources.py` file. a. if we move the DummyRados definition to the test file, without removing the usage in the `create-external-cluster-resources.py` file we will create a cyclic dependency. b. if we remove all the references of DummyRados from the `create-external-cluster-resources.py` file and then move it to the test file, then it should be fine, but we will face difficulties in running the test.
> > > The clean way out will be to take DummyRados class out to it's own `py` file and import it accordingly.
> > > I agree to @parth-gr in the sentiments that we should not include/introduce DummyRados into the actual code. But we need to find a workaround for it...
> > 
> > 
> > are we good for this pr or do we need to think about some workaround and include it in this PR?
> > > Thanks
> 
> Changing the `DummyRados` (IMHO) to a different file and changing the references can be taken up as a different refactoring PR. But I've a small concern regarding `importlib` package, which is only there in 3.x python version. Is there any way we can make this backward compatible?
> 
> If are not concerned about the 2.7.x python versions, then LGTM from my end...

Thanks, @aruniiird for the point we discussed in the huddle that it is just unit test so we should be good with backward compatible

**Pull request merged**